crates/flux-refineck/src/checker.rs

```diff
@@ -315,13 +315,15 @@ fn check_fn_subtyping(
     infcx
         .subtyping(&output.ret, &super_output.ret, reason)
         .with_span(span)?;
+    let evars_sol = infcx.pop_scope().with_span(span)?;
```

Comment:
These three lines are the fix - my q. is why isn't it enough to just have the stuff at the bottom? (the very last infcx.replace_evars) I note that the `update_ensures` calls `infcx.unpack` and that somehow messes stuff up and necessitates the `replace_evars` call BEFORE `update_ensures`, but why?

Reply:
The `InferCtxt` is a cursor into a tree. Some functions like `define_var`, `assume_pred` and by extension `unpack` _advance_ the cursor. For example, defining a variable pushes a node as a child of the current node and then moves the cursor into that new node. Other functions, like `subtyping` or  `check_pred` (typically the ones defined in `InferCtxtAt`) _do not_ advance the cursor (from the caller's perspective). For example, if you call `subtyping` a subtree is pushed under the current node, and the cursor is returned to where it was. That's the purpose of `infcx.branch`: to "clone" the cursor such that the original one doesn't get modified.

`infcx.pop_scope` and `infcx.replace_evars` are supposed to be called when the `infcx` is on the same node where the `push_scope` was called, because `replace_evars` only replaces evars below that node.  What was happening here was that `pop_scope` was being called at a node further down (after calling `unpack`) so evars above that weren't being replaced.

This is very error-prone. The whole evar infrastructure needs some revamping.

Reply:
We could accumulate solutions and apply them at the very end. The only evars we need to solve eagerly are the ones for locations because we need them to apply the ensures. Locations aside, we could even generate the constraint with evars and then solve them _in_ the constraint.

Reply:
I see -- definitely quite delicate. I suggest we merge this in right now, though, (assuming that the push/pop is now done properly) and come back to clean up this push/pop/evar API later.